Flow control proxy and tests

mfcripps opened the pull request:
no rate limiting or delay yet.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+    qLength = bandwidth * latency / 1000; // convert from ms to s
+    chunkSize = Math.min(qLength, 1024);
+
+    System.out.println("Started new proxy on port " + proxyPort);
```
make this a logger statement

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  public void setUp() throws IOException {
+    qLength = bandwidth * latency / 1000; // convert from ms to s
```
can qLength be 0?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  private synchronized void incrementServerQSize(int amt) {
```
amount

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+      client.close();
+    } catch (IOException e) {
+      System.out.println("Shutdown failed");
+      e.printStackTrace();
```
just allow the exception to propagate?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+    DataInputStream serverIn = new DataInputStream(server.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    (new Thread(new ClientReader(clientIn))).start();
```
How will these threads be shutdown?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+    byte[] message;
+    int messageLength;
+
+    public Message(long at, byte[] m, int ml) {
```
please use slightly longer names.   It isn't clear from the types or the context what these are.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+          Long howLong = (next.arrivalTime + latency) - System.currentTimeMillis();
+          while (howLong > 0) {
+            System.out.println("how long: " + howLong);
+            Thread.sleep(howLong);
```
Can you use the sleep overload that accepts nanos?  Or possibly use TimeUnit.sleep?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+
+    @Override
+    public void run() {
+      while (true) {
```
You should probably make some sort of sentinel variable that this checks.  When you want to shut down, this thread will exit the loop.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  private class ServerWriter implements Runnable {
```
Alternatively you could just extend Thread

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+      while (true) {
+        try {
+          while (clientQueue.isEmpty()) {
+            Thread.sleep(1);
```
You can use a LinkedBlockingQueue and not busy-loop awaiting input.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+          incrementServerQSize(readableResponseBytes);
+          System.out.println("Server q size: " + serverQSize);
+        } catch (IOException e) {
+          e.printStackTrace();
```
Just let these exceptions propagate and set an Uncaught exception handler on the thread.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+        try {
+          byte[] response = new byte[chunkSize];
+          int readableResponseBytes = serverIn.read(response);
+          long arrivalTime = System.currentTimeMillis();
```
currentTimeMillis is not useful for measuring time here, since it can go backwards.  use System.nanoTime instead.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,248 @@
+          long arrivalTime = System.currentTimeMillis();
+          Message m = new Message(arrivalTime, request, readableRequestBytes);
+          while (clientQSize + readableRequestBytes > qLength) {
+            System.out.println(clientQSize + readableRequestBytes);
```
This should also be a log.

mfcripps commented:
PTAL

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  private class ClientReader implements Runnable {
```
I was surprised to see separate implementations for client and server. It looks like only the queue to use is different? Maybe pass in `serverQueue`/`clientQueue` as a constructor parameter?

ejona86 replied:
Oh, I see there is `incrementClientQSize` as well. Maybe there should be an additional class that has state for a particular direction and the Runnables?

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  private class Message {
```
nit: static. Just makes it easier to read.

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,231 @@
+          Message next = serverQueue.take();
+          long howLong = (next.arrivalTime + latency) - (System.nanoTime() / 1000000);
+          while (howLong > 0) {
+            Thread.sleep(howLong);
```
You actually have to do the sleeping on the reader, not the writer. Otherwise you re-produce the issues with infinitely buffering. By avoiding reading on the client, you cause push-back on the sender very similar to if there were lost packets due to going too fast.

ejona86 replied:
Oh, this is okay. I see now how the client sleeps for 1ms at a time when the queue is "full".

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,231 @@
+public class TrafficControlProxy {
+
+  private static int DEFAULT_BAND = 1024 * 1024; // Bps
+  private static int DEFAULT_DELAY = 100; // ms;
```
nit: put milliseconds in the name, so it is obvious when being used that it is being used correctly, so DEFAULT_DELAY_MS.

mfcripps commented:
@carl-mastrangelo PTAL

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.logging.Logger;
+
+public class TrafficControlProxy {
```
Please make this final

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private static int DEFAULT_DELAY_MS = 100; // ms;
+  private static Logger logger = Logger.getLogger(TrafficControlProxy.class.getName());
+
+  private String loopback = "127.127.127.127";
```
Since you aren't using tc, you should probably use 127.0.0.1

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  // TODO: make ports arguments
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int qLength;
```
queueLength

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int qLength;
+  private int serverQSize;
```
and here and below

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private int clientQSize;
+  private int chunkSize;
+  private boolean shutDown;
+  private LinkedBlockingQueue<Message> serverQueue = new LinkedBlockingQueue<Message>();
```
final

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private Socket server;
+  private Socket clientSock;
+
+
```
nit: extra space

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private int latency = DEFAULT_DELAY_MS;
+
+  public void setBandwidth(int targetBPS) {
+    bandwidth = targetBPS;
```
checkArgument that bandwidth is positive?

carl-mastrangelo replied:
If you don't plan on allowing this value to change once started, put it in the constructor.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+
+  private int bandwidth = DEFAULT_BAND;
+  private int latency = DEFAULT_DELAY_MS;
+
```
This method probably deserves a javadoc since it is public.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+
+
+  public void setLatency(int targetMillis) {
+    latency = targetMillis / 2; // divide by two because delay is applied in both directions
```
same comment.  Always sanity check input values on public methods.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  }
+
+
+  public void setLatency(int targetMillis) {
```
This should probably be a TimeUnit and a value, like we do elsewhere in our code.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  public void setUp() throws IOException {
+    qLength = Math.max(bandwidth * latency / 1000, 1); // convert from ms to s
```
Again, let TimeUnit do the conversion.   Timeunit.FOO.toSeconds(....)

mfcripps replied:
Converting from smaller time units to larger ones truncates the value instead of returning a fractional result, so doesn't seem to work for what this use cases. Should be fine for the other 'howLong' sleeps though.  Any other suggestions for this besides regular division?

carl-mastrangelo replied:
Use nanoseconds throughout your whole program.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+
+  public void setUp() throws IOException {
+    qLength = Math.max(bandwidth * latency / 1000, 1); // convert from ms to s
+    chunkSize = Math.min(qLength, 4 * 1024);
```
Why 4k?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    chunkSize = Math.min(qLength, 4 * 1024);
+
+    logger.info("Started new proxy on port " + proxyPort);
+    client = new ServerSocket();
```
client uses server socket?  and server uses Socket?  Might be worth a comment, since there are two clients and two servers in a proxy.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    logger.info("Started new proxy on port " + proxyPort);
+    client = new ServerSocket();
+    client.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = client.accept();
```
Won't this be a oneshot proxy?   If you try and reuse this it will fail right?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    client.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = client.accept();
+    server = new Socket();
+    server.connect(new InetSocketAddress(loopback, serverPort));
```
terminology is going to be important here.  The "server" could be the proxy itself, since it accepts connections from other endpoints.  The server could also mean the destination that this proxy forwards to.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    DataInputStream serverIn = new DataInputStream(server.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+
```
nit: extra space.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  private int serverQSize;
+  private int clientQSize;
+  private int chunkSize;
+  private boolean shutDown;
```
This is accessed from two different threads, so it needs to be properly synchronized.  You can do this by making it volatile, or adding it in a synchronized block.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  private synchronized void incrementServerQSize(int delta) {
+    serverQSize += delta;
```
This is also improperly synchronized.  the run() method below reads from serverQSize, but not in a synchronized block.  It may never find out that another thread changed this value.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  private synchronized void incrementClientQSize(int delta) {
+    clientQSize += delta;
```
Same here.  You run risks of calling this method and it strangely not working.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+
+  private class ClientReader implements Runnable {
+
+    DataInputStream clientIn;
```
private final

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    DataInputStream clientIn;
+
+    public ClientReader(DataInputStream clientIn) {
+      this.clientIn = clientIn;
```
this.clientIn = checkNotNull(clientIn);

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+
+    DataInputStream clientIn;
+
+    public ClientReader(DataInputStream clientIn) {
```
no need to specify public

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    p.setUp();
+  }
+
+  private class ClientReader implements Runnable {
```
final

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = Math.min(chunkSize, qLength - serverQSize);
```
can this return a negative value?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+        try {
+          int nextChunk = Math.min(chunkSize, qLength - serverQSize);
+          if (nextChunk != 0) {
+            nextChunk = Math.min(chunkSize, qLength - serverQSize);
```
why is this duplicated from the line above?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+            nextChunk = Math.min(chunkSize, qLength - serverQSize);
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = clientIn.read(request);
+            long arrivalTime = System.nanoTime() / 1000000;
```
units?  again, TimeUnit is your friend here.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+            int readableRequestBytes = clientIn.read(request);
+            long arrivalTime = System.nanoTime() / 1000000;
+            Message m = new Message(arrivalTime, request, readableRequestBytes);
+            clientQueue.add(m);
```
just inline the add()

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+            incrementClientQSize(readableRequestBytes);
+          } else {
+            // This is arbitrary right now. Maybe sleep for some fraction of the latency?
+            Thread.sleep(1);
```
why sleep?

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+            Thread.sleep(1);
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
```
getting an exception should probably end the proxy.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
```
InterruptedException is special.  You need to reset the interrupted bit if you aren't going to handle it.  Printing it is not a good way of handling it.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,245 @@
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = Math.min(chunkSize, qLength - serverQSize);
```
Why do both ClientReader and ServerReader use the server queue size?

carl-mastrangelo commented on interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java:
```diff
@@ -0,0 +1,230 @@
+      clientIn.read(message);
+    }
+    // test
+    long start = System.currentTimeMillis();
```
don't use currentTimeMillis.  It can do funky things like go backwards in time.  Use nanoTime unless you know that cTM is really what you want.

carl-mastrangelo commented on interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java:
```diff
@@ -0,0 +1,230 @@
+
+  @Test
+  @Ignore
+  public void smallBandwidth() throws UnknownHostException, IOException, InterruptedException {
```
just throw Exception here in tests

mfcripps commented:
PTAL. Tried to make most of the changes from the earlier commit, but there are still issues with synchronization and exception handling. I held off on some of the exception handling/shut down issues because I think it will change as the threading model changes.  So right now, this is still a one shot proxy that doesn't shut down cleanly.

mfcripps commented:
accidentally pushed FlowControlTest.java, please ignore

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,256 @@
+  private int chunkSize;
+  private int bandwidth;
+  private long latency;
+  private final LinkedBlockingQueue<Message> serverQueue = new LinkedBlockingQueue<Message>();
```
Just use BlockingQueue as the assigned type.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,256 @@
+   * @param latencyUnits
+   */
+  public TrafficControlProxy(int targetBPS, int targetLatency, TimeUnit latencyUnits) {
+    Preconditions.checkArgument(targetBPS > 0);
```
FYI you can statically import checkArgument.   In Eclipse you can select it and pres Ctrl Shift M.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,256 @@
+    Preconditions.checkArgument(targetBPS > 0);
+    Preconditions.checkArgument(targetLatency > 0);
+    bandwidth = targetBPS;
+    latency = latencyUnits.toNanos(targetLatency) / 2; // Divide by 2 because delay is applied in
```
Only inline a comment if it will fit on the same line.  Otherwise they go above.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,256 @@
+            incrementClientQueueSize(readableRequestBytes);
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
```
Just rethrow the exception here as new RuntimeException(e);

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,256 @@
+            incrementServerQueueSize(readableResponseBytes);
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
```
And here and below

carl-mastrangelo commented on interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java:
```diff
@@ -0,0 +1,233 @@
+  // easier).
+
+  @Test
+  @Ignore
```
You can add a comment to ignores I thinkg:

`@Ignore("ignored until lorem ipsum")`

carl-mastrangelo commented on interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java:
```diff
@@ -0,0 +1,233 @@
+  @Ignore
+  public void bigLatency() throws UnknownHostException, IOException, InterruptedException {
+    Server server = new Server();
+    (new Thread(server)).start();
```
FYI threads can be started, but they can't be explicitly ended from the outside.   What you can do is wait for them to finish.  This is called joining.  You should keep a reference to the thread, and call thread.join() from this thread to make sure it finishes.

carl-mastrangelo commented:
@mfcripps Added some more comments.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+// import java.lang.annotation.Native;
```
remove

mfcripps replied:
Removed all unused imports

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+  private long latency;
+  private final BlockingQueue<Message> serverQueue = new DelayQueue<Message>();
+  private final BlockingQueue<Message> clientQueue = new DelayQueue<Message>();
+  private volatile int serverQueueSize;
```
I don't think `serverQueueSize` needs to be volatile since you always access it under a lock, or after a thread is exited.

mfcripps replied:
removed

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -220,37 +284,36 @@ public void run() {
-    public String toString() {
-      return "" + messageLength;
+    public int compareTo(Delayed o) {
+      return ((Long) getDelay(TimeUnit.NANOSECONDS)).compareTo(o.getDelay(TimeUnit.NANOSECONDS));
```
Compare `sendTime` directly? You can unconditionally cast to `Message`.

mfcripps replied:
changed to compare sendTime

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+  private final BlockingQueue<Message> serverQueue = new DelayQueue<Message>();
+  private final BlockingQueue<Message> clientQueue = new DelayQueue<Message>();
+  private volatile int serverQueueSize;
+  private volatile int clientQueueSize;
```
same for this

mfcripps replied:
also removed

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    clientAcceptor.close();
+    clientSock.close();
```
nit: if one of these fails, the subsequent closes won't work.

Maybe add a TODO to hand this case?

mfcripps replied:
Haven't handled this yet but added a TODO

ejona86 commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -63,63 +72,125 @@ public TrafficControlProxy() {
+    while (serverQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
```
This should actually propagate out of this method. By suppressing the interruption here, it seems like readers could block forever via a race with `shutdown` being set to `true`.

mfcripps replied:
Switched these methods to allow Interrupted exception to propagate

ejona86 commented:
#1 feature request: shared class used for both client and server :-D

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+            byte[] request = new byte[nextChunk];
+            int readableRequestBytes = serverIn.read(request);
+            long sendTime = System.nanoTime() + latency;
+            serverQueue.offer(new Message(sendTime, request, readableRequestBytes));
```
Offer returns a boolean saying if it worked.  I think you either need to check this value, or use the blocking `take` method.

mfcripps replied:
Swapped to put, which should always succeed on a delay queue because delay queues are unbounded.

carl-mastrangelo commented on interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,319 @@
+      while (!shutDown) {
+        try {
+          int nextChunk = getServerChunk();
+          if (nextChunk != 0) {
```
Why not just let the exception propagate?   You always interrupt when you are done with work, so there isn't any chance of recovery.

mfcripps replied:
Makes sense, took out this check.

mfcripps commented:
PTAL

ejona86 commented on interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  private void incrementQueue(Endpoint endpoint, int delta) {
+    if (endpoint == Endpoint.SERVER) {
```
Instead of an enum, instead just put serverQueueSize, serverQueue, etc, in a class along with these methods and the Runnables.

mfcripps replied:
Put all methods that interact with the DelayQueue in a new class MessageQueue, but I left the Reader and Writer runnables as their own classes that take a MessageQueue as an argument.

mfcripps commented:
PTAL

ejona86 commented on interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,273 @@
+    clientAcceptor.close();
+    clientSock.close();
+    serverSock.close();
+    for (Thread worker : workers) {
```
Do another loop over the workers and join() on them? It seems like `reset()` requires all the threads to be exited by the time this method returns.

mfcripps replied:
makes sense, added a loop to call join

ejona86 commented on interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java:
```diff
@@ -0,0 +1,273 @@
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+    workers.clear();
```
Maybe clear the workers in `shutDown()`, since it seems you allow shutting down and then starting again?

mfcripps replied:
moved clear to shutdown

ejona86 commented:
Some minor comments, but LGTM. That separate class makes it much nicer.

mfcripps commented:
Modified to work with integration tests - 
Now uses a separate thread for accepting client/server connections. To make this easier I switched over from a worker list to a ThreadPoolExecutor.
